src/index.ts

```diff
@@ -95,41 +96,26 @@ const calculateDepreciation = ({
     }]
   }
 
-  for (let index = 0; index < totalDepreciationYears; index++) {
-    // Current year
-    if (index === 0) {
-      const monthsLeftInFirstYear = MONTHS_IN_YEAR - purchaseMonth + 1;
-      const result = calculate(purchaseAmount, totalDepreciationYears, endAmount, monthsLeftInFirstYear);
-      results.push({
-        year: purchaseYear,
-        depreciationMonths: monthsLeftInFirstYear,
-        ...result,
-      });
-      endAmount = result.endAmount;
-
-      // If item is purchased in Jan, monthsLeftInLastYear is 0
-      // If item is purchased in Dec, monthsLeftInLastYear is 11
-      monthsLeftInLastYear = MONTHS_IN_YEAR - monthsLeftInFirstYear;
-    } else { // Future years
-      const result = calculate(purchaseAmount, totalDepreciationYears, endAmount);
-      results.push({
-        year: purchaseYear + index,
-        depreciationMonths: MONTHS_IN_YEAR,
-        ...result,
-      });
-      endAmount = result.endAmount;
-    }
-  }
+  const parts: number = purchaseMonth > 1 ? totalDepreciationYears + 1 : totalDepreciationYears;
+  const monthInFirstYear : number = MONTHS_IN_YEAR - purchaseMonth + 1;
+  const monthInLastYear : number = monthInFirstYear === MONTHS_IN_YEAR ? MONTHS_IN_YEAR : MONTHS_IN_YEAR - monthInFirstYear;
 
-  if (monthsLeftInLastYear > 0) {
+  let monthInThisYear : number = monthInFirstYear;
+  endAmount = purchaseAmount;
+
+  for (let index = 0; index < parts; index++) {        
+    const result = calculate(purchaseAmount, totalDepreciationYears, endAmount, monthInThisYear, index === parts - 1);
     results.push({
-      year: purchaseYear + totalDepreciationYears,
-      depreciationMonths: monthsLeftInLastYear,
-      ...calculate(purchaseAmount, totalDepreciationYears, endAmount, monthsLeftInLastYear),
+      year: purchaseYear + index,
+      depreciationMonths: monthInThisYear,
+      ...result,
     });
+    endAmount = result.endAmount;  
+    // calculate month in year for next step and control the last year case.
+    monthInThisYear = (index + 1 === parts - 1) ? monthInLastYear : MONTHS_IN_YEAR;
```

Comment:
Because I didn't want this line to effect on first year, I moved this line to end of loop, I agree with you that was added some complexity to code and I also didn't want to remove the `calculate` function, I think this function may be useful in future extensions of code as well as help to better understand of code, so for easier understanding I made some changes and remove this line by calculation the months in each year and storing them on an array.